File: performance-benchmarks/python-implementation/temporal_tigerbeetle_integration.py
"""
Python implementation of Temporal-TigerBeetle integration for performance comparison.
This implementation mirrors the Go version to enable fair benchmarking.
"""

import asyncio
import time
import uuid
from dataclasses import dataclass
from datetime import datetime, timedelta
from enum import Enum
from typing import Optional, List, Dict, Any

from temporalio import activity, workflow
from temporalio.client import Client
from temporalio.worker import Worker


class TransferStatus(Enum):
    PENDING = "pending"
    COMMITTED = "committed"
    VOIDED = "voided"
    FAILED = "failed"


@dataclass
class TransferRequest:
    transfer_id: str
    debit_account_id: str
    credit_account_id: str
    amount: int
    ledger: int
    code: int
    timeout: int = 3600
    is_pending: bool = False


@dataclass
class TransferResult:
    transfer_id: str
    status: str
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    timestamp: datetime = None

    def __post_init__(self):
        if self.timestamp is None:
            self.timestamp = datetime.utcnow()


@dataclass
class AccountBalance:
    account_id: str
    debits_posted: int
    credits_posted: int
    debits_pending: int
    credits_pending: int
    net_balance: int


class TigerBeetleClient:
    """
    Python TigerBeetle client wrapper.
    Note: This is a simulated client for benchmarking purposes.
    In production, use the official tigerbeetle-python client.
    """

    def __init__(self, cluster_id: int, addresses: List[str]):
        self.cluster_id = cluster_id
        self.addresses = addresses
        self._accounts: Dict[str, Dict[str, int]] = {}
        self._transfers: Dict[str, TransferResult] = {}
        self._lock = asyncio.Lock()

    async def create_account(
        self, account_id: str, ledger: int, code: int
    ) -> None:
        async with self._lock:
            if account_id in self._accounts:
                raise ValueError(f"Account {account_id} already exists")
            
            self._accounts[account_id] = {
                "ledger": ledger,
                "code": code,
                "debits_posted": 0,
                "credits_posted": 0,
                "debits_pending": 0,
                "credits_pending": 0,
            }
            
            # Simulate network latency
            await asyncio.sleep(0.001)

    async def create_transfer(self, req: TransferRequest) -> TransferResult:
        async with self._lock:
            if req.debit_account_id not in self._accounts:
                return TransferResult(
                    transfer_id=req.transfer_id,
                    status="failed",
                    error_code="DEBIT_ACCOUNT_NOT_FOUND",
                    error_message=f"Debit account {req.debit_account_id} not found",
                )

            if req.credit_account_id not in self._accounts:
                return TransferResult(
                    transfer_id=req.transfer_id,
                    status="failed",
                    error_code="CREDIT_ACCOUNT_NOT_FOUND",
                    error_message=f"Credit account {req.credit_account_id} not found",
                )

            debit_account = self._accounts[req.debit_account_id]
            credit_account = self._accounts[req.credit_account_id]

            if req.is_pending:
                debit_account["debits_pending"] += req.amount
                credit_account["credits_pending"] += req.amount
                status = "pending"
            else:
                available_balance = (
                    debit_account["credits_posted"] - debit_account["debits_posted"]
                )
                if available_balance < req.amount:
                    return TransferResult(
                        transfer_id=req.transfer_id,
                        status="failed",
                        error_code="INSUFFICIENT_BALANCE",
                        error_message="Insufficient balance",
                    )

                debit_account["debits_posted"] += req.amount
                credit_account["credits_posted"] += req.amount
                status = "committed"

            self._transfers[req.transfer_id] = TransferResult(
                transfer_id=req.transfer_id,
                status=status,
            )

            # Simulate network latency
            await asyncio.sleep(0.002)

            return self._transfers[req.transfer_id]

    async def post_pending_transfer(
        self, transfer_id: str, pending_transfer_id: str, ledger: int, code: int
    ) -> TransferResult:
        async with self._lock:
            if pending_transfer_id not in self._transfers:
                return TransferResult(
                    transfer_id=transfer_id,
                    status="failed",
                    error_code="PENDING_TRANSFER_NOT_FOUND",
                    error_message=f"Pending transfer {pending_transfer_id} not found",
                )

            pending_transfer = self._transfers[pending_transfer_id]
            if pending_transfer.status != "pending":
                return TransferResult(
                    transfer_id=transfer_id,
                    status="failed",
                    error_code="TRANSFER_NOT_PENDING",
                    error_message="Transfer is not in pending state",
                )

            # Find the original transfer details (simplified)
            # In real implementation, we'd store transfer details
            pending_transfer.status = "committed"

            result = TransferResult(
                transfer_id=transfer_id,
                status="committed",
            )

            self._transfers[transfer_id] = result

            # Simulate network latency
            await asyncio.sleep(0.002)

            return result

    async def void_pending_transfer(
        self, transfer_id: str, pending_transfer_id: str, ledger: int, code: int
    ) -> TransferResult:
        async with self._lock:
            if pending_transfer_id not in self._transfers:
                return TransferResult(
                    transfer_id=transfer_id,
                    status="failed",
                    error_code="PENDING_TRANSFER_NOT_FOUND",
                    error_message=f"Pending transfer {pending_transfer_id} not found",
                )

            pending_transfer = self._transfers[pending_transfer_id]
            if pending_transfer.status != "pending":
                return TransferResult(
                    transfer_id=transfer_id,
                    status="failed",
                    error_code="TRANSFER_NOT_PENDING",
                    error_message="Transfer is not in pending state",
                )

            pending_transfer.status = "voided"

            result = TransferResult(
                transfer_id=transfer_id,
                status="voided",
            )

            self._transfers[transfer_id] = result

            # Simulate network latency
            await asyncio.sleep(0.002)

            return result

    async def get_account_balance(self, account_id: str) -> AccountBalance:
        async with self._lock:
            if account_id not in self._accounts:
                raise ValueError(f"Account {account_id} not found")

            account = self._accounts[account_id]
            net_balance = account["credits_posted"] - account["debits_posted"]

            # Simulate network latency
            await asyncio.sleep(0.001)

            return AccountBalance(
                account_id=account_id,
                debits_posted=account["debits_posted"],
                credits_posted=account["credits_posted"],
                debits_pending=account["debits_pending"],
                credits_pending=account["credits_pending"],
                net_balance=net_balance,
            )

    def close(self):
        pass


class TigerBeetleActivities:
    def __init__(self, client: TigerBeetleClient):
        self.client = client

    @activity.defn(name="CreateAccountActivity")
    async def create_account_activity(
        self, account_id: str, ledger: int, code: int
    ) -> None:
        activity.logger.info(f"Creating TigerBeetle account: {account_id}")
        await self.client.create_account(account_id, ledger, code)
        activity.logger.info(f"Account created successfully: {account_id}")

    @activity.defn(name="CreateTransferActivity")
    async def create_transfer_activity(self, req: TransferRequest) -> TransferResult:
        activity.logger.info(f"Creating TigerBeetle transfer: {req.transfer_id}")
        result = await self.client.create_transfer(req)
        activity.logger.info(
            f"Transfer created: {req.transfer_id}, status: {result.status}"
        )
        return result

    @activity.defn(name="PostPendingTransferActivity")
    async def post_pending_transfer_activity(
        self, transfer_id: str, pending_transfer_id: str, ledger: int, code: int
    ) -> TransferResult:
        activity.logger.info(
            f"Posting pending transfer: {transfer_id}, pending: {pending_transfer_id}"
        )
        result = await self.client.post_pending_transfer(
            transfer_id, pending_transfer_id, ledger, code
        )
        activity.logger.info(f"Pending transfer posted: {transfer_id}")
        return result

    @activity.defn(name="VoidPendingTransferActivity")
    async def void_pending_transfer_activity(
        self, transfer_id: str, pending_transfer_id: str, ledger: int, code: int
    ) -> TransferResult:
        activity.logger.info(
            f"Voiding pending transfer: {transfer_id}, pending: {pending_transfer_id}"
        )
        result = await self.client.void_pending_transfer(
            transfer_id, pending_transfer_id, ledger, code
        )
        activity.logger.info(f"Pending transfer voided: {transfer_id}")
        return result

    @activity.defn(name="GetAccountBalanceActivity")
    async def get_account_balance_activity(self, account_id: str) -> AccountBalance:
        activity.logger.info(f"Getting account balance: {account_id}")
        balance = await self.client.get_account_balance(account_id)
        activity.logger.info(
            f"Account balance retrieved: {account_id}, net: {balance.net_balance}"
        )
        return balance

    @activity.defn(name="ValidatePaymentActivity")
    async def validate_payment_activity(self, payment_id: str) -> bool:
        # Simulate payment validation
        await asyncio.sleep(0.01)
        return True


@dataclass
class PaymentWorkflowInput:
    payment_id: str
    debit_account_id: str
    credit_account_id: str
    amount: int
    currency: str
    ledger: int
    code: int


@dataclass
class PaymentWorkflowResult:
    payment_id: str
    transfer_id: str
    status: str
    error_message: Optional[str] = None
    completed_at: datetime = None

    def __post_init__(self):
        if self.completed_at is None:
            self.completed_at = datetime.utcnow()


@workflow.defn(name="PaymentWorkflow")
class PaymentWorkflow:
    @workflow.run
    async def run(self, input: PaymentWorkflowInput) -> PaymentWorkflowResult:
        workflow.logger.info(f"Starting payment workflow: {input.payment_id}")

        transfer_id = f"TXN-{input.payment_id}-{int(time.time())}"

        # Create pending transfer
        pending_req = TransferRequest(
            transfer_id=transfer_id,
            debit_account_id=input.debit_account_id,
            credit_account_id=input.credit_account_id,
            amount=input.amount,
            ledger=input.ledger,
            code=input.code,
            timeout=3600,
            is_pending=True,
        )

        pending_result = await workflow.execute_activity(
            "CreateTransferActivity",
            pending_req,
            start_to_close_timeout=timedelta(seconds=30),
        )

        if pending_result.status == "failed":
            workflow.logger.error(
                f"Failed to create pending transfer: {pending_result.error_message}"
            )
            return PaymentWorkflowResult(
                payment_id=input.payment_id,
                transfer_id=transfer_id,
                status="failed",
                error_message=pending_result.error_message,
            )

        # Validate payment
        try:
            payment_approved = await workflow.execute_activity(
                "ValidatePaymentActivity",
                input.payment_id,
                start_to_close_timeout=timedelta(seconds=30),
            )
        except Exception as e:
            workflow.logger.error(f"Payment validation failed: {e}")
            
            # Void the pending transfer
            void_id = f"VOID-{transfer_id}"
            await workflow.execute_activity(
                "VoidPendingTransferActivity",
                void_id,
                transfer_id,
                input.ledger,
                input.code,
                start_to_close_timeout=timedelta(seconds=30),
            )
            
            return PaymentWorkflowResult(
                payment_id=input.payment_id,
                transfer_id=transfer_id,
                status="failed",
                error_message="Payment validation failed",
            )

        if not payment_approved:
            workflow.logger.info("Payment not approved, voiding transfer")
            
            void_id = f"VOID-{transfer_id}"
            await workflow.execute_activity(
                "VoidPendingTransferActivity",
                void_id,
                transfer_id,
                input.ledger,
                input.code,
                start_to_close_timeout=timedelta(seconds=30),
            )
            
            return PaymentWorkflowResult(
                payment_id=input.payment_id,
                transfer_id=transfer_id,
                status="rejected",
                error_message="Payment not approved",
            )

        # Post pending transfer
        post_id = f"POST-{transfer_id}"
        post_result = await workflow.execute_activity(
            "PostPendingTransferActivity",
            post_id,
            transfer_id,
            input.ledger,
            input.code,
            start_to_close_timeout=timedelta(seconds=30),
        )

        if post_result.status == "failed":
            workflow.logger.error(
                f"Failed to post pending transfer: {post_result.error_message}"
            )
            return PaymentWorkflowResult(
                payment_id=input.payment_id,
                transfer_id=transfer_id,
                status="failed",
                error_message=post_result.error_message,
            )

        workflow.logger.info(f"Payment workflow completed: {input.payment_id}")
        return PaymentWorkflowResult(
            payment_id=input.payment_id,
            transfer_id=transfer_id,
            status="completed",
        )


async def create_worker(
    temporal_host: str,
    temporal_namespace: str,
    task_queue: str,
    tigerbeetle_addresses: List[str],
) -> Worker:
    """Create and configure a Temporal worker with TigerBeetle activities."""
    
    client = await Client.connect(temporal_host, namespace=temporal_namespace)
    
    tigerbeetle_client = TigerBeetleClient(
        cluster_id=0,
        addresses=tigerbeetle_addresses,
    )
    
    activities_instance = TigerBeetleActivities(tigerbeetle_client)
    
    worker = Worker(
        client,
        task_queue=task_queue,
        workflows=[PaymentWorkflow],
        activities=[
            activities_instance.create_account_activity,
            activities_instance.create_transfer_activity,
            activities_instance.post_pending_transfer_activity,
            activities_instance.void_pending_transfer_activity,
            activities_instance.get_account_balance_activity,
            activities_instance.validate_payment_activity,
        ],
        max_concurrent_workflow_tasks=100,
        max_concurrent_activities=200,
    )
    
    return worker, tigerbeetle_client

File: bidirectional-integrations/temporal-tigerbeetle/temporal_tigerbeetle_integration.go
package temporaltigerbeetle

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/workflow"
	tb "github.com/tigerbeetle/tigerbeetle-go"
	"github.com/tigerbeetle/tigerbeetle-go/pkg/types"
	"github.com/google/uuid"
)

type TigerBeetleClient struct {
	client tb.Client
}

type TransferRequest struct {
	TransferID      string
	DebitAccountID  types.Uint128
	CreditAccountID types.Uint128
	Amount          uint64
	Ledger          uint32
	Code            uint16
	Timeout         uint32
	IsPending       bool
}

type TransferResult struct {
	TransferID   string
	Status       string
	ErrorCode    string
	ErrorMessage string
	Timestamp    time.Time
}

type AccountBalance struct {
	AccountID      types.Uint128
	DebitsPosted   uint64
	CreditsPosted  uint64
	DebitsPending  uint64
	CreditsPending uint64
	NetBalance     int64
}

func NewTigerBeetleClient(clusterID types.Uint128, addresses []string) (*TigerBeetleClient, error) {
	client, err := tb.NewClient(clusterID, addresses)
	if err != nil {
		return nil, fmt.Errorf("failed to create TigerBeetle client: %w", err)
	}

	return &TigerBeetleClient{
		client: client,
	}, nil
}

func (c *TigerBeetleClient) CreateAccount(ctx context.Context, accountID types.Uint128, ledger uint32, code uint16) error {
	accounts := []types.Account{
		{
			ID:        accountID,
			Ledger:    ledger,
			Code:      code,
			Flags:     0,
			Timestamp: uint64(time.Now().UnixNano()),
		},
	}

	results, err := c.client.CreateAccounts(accounts)
	if err != nil {
		return fmt.Errorf("failed to create account: %w", err)
	}

	if len(results) > 0 {
		return fmt.Errorf("account creation failed with result: %v", results[0].Result)
	}

	return nil
}

func (c *TigerBeetleClient) CreateTransfer(ctx context.Context, req TransferRequest) (*TransferResult, error) {
	transferID := parseTransferID(req.TransferID)

	var flags uint16
	if req.IsPending {
		flags = types.TransferFlags{Pending: true}.ToUint16()
	}

	transfers := []types.Transfer{
		{
			ID:              transferID,
			DebitAccountID:  req.DebitAccountID,
			CreditAccountID: req.CreditAccountID,
			Amount:          req.Amount,
			Ledger:          req.Ledger,
			Code:            req.Code,
			Flags:           flags,
			Timeout:         req.Timeout,
			Timestamp:       uint64(time.Now().UnixNano()),
		},
	}

	results, err := c.client.CreateTransfers(transfers)
	if err != nil {
		return &TransferResult{
			TransferID:   req.TransferID,
			Status:       "failed",
			ErrorMessage: err.Error(),
			Timestamp:    time.Now(),
		}, fmt.Errorf("failed to create transfer: %w", err)
	}

	if len(results) > 0 {
		return &TransferResult{
			TransferID:   req.TransferID,
			Status:       "failed",
			ErrorCode:    fmt.Sprintf("%d", results[0].Result),
			ErrorMessage: "Transfer creation failed",
			Timestamp:    time.Now(),
		}, fmt.Errorf("transfer creation failed with result: %v", results[0].Result)
	}

	return &TransferResult{
		TransferID: req.TransferID,
		Status:     "success",
		Timestamp:  time.Now(),
	}, nil
}

func (c *TigerBeetleClient) PostPendingTransfer(ctx context.Context, transferID, pendingTransferID string, ledger uint32, code uint16) (*TransferResult, error) {
	postID := parseTransferID(transferID)
	pendingID := parseTransferID(pendingTransferID)

	transfers := []types.Transfer{
		{
			ID:        postID,
			PendingID: pendingID,
			Ledger:    ledger,
			Code:      code,
			Flags:     types.TransferFlags{PostPendingTransfer: true}.ToUint16(),
			Timestamp: uint64(time.Now().UnixNano()),
		},
	}

	results, err := c.client.CreateTransfers(transfers)
	if err != nil {
		return &TransferResult{
			TransferID:   transferID,
			Status:       "failed",
			ErrorMessage: err.Error(),
			Timestamp:    time.Now(),
		}, fmt.Errorf("failed to post pending transfer: %w", err)
	}

	if len(results) > 0 {
		return &TransferResult{
			TransferID:   transferID,
			Status:       "failed",
			ErrorCode:    fmt.Sprintf("%d", results[0].Result),
			ErrorMessage: "Post pending transfer failed",
			Timestamp:    time.Now(),
		}, fmt.Errorf("post pending transfer failed with result: %v", results[0].Result)
	}

	return &TransferResult{
		TransferID: transferID,
		Status:     "committed",
		Timestamp:  time.Now(),
	}, nil
}

func (c *TigerBeetleClient) VoidPendingTransfer(ctx context.Context, transferID, pendingTransferID string, ledger uint32, code uint16) (*TransferResult, error) {
	voidID := parseTransferID(transferID)
	pendingID := parseTransferID(pendingTransferID)

	transfers := []types.Transfer{
		{
			ID:        voidID,
			PendingID: pendingID,
			Ledger:    ledger,
			Code:      code,
			Flags:     types.TransferFlags{VoidPendingTransfer: true}.ToUint16(),
			Timestamp: uint64(time.Now().UnixNano()),
		},
	}

	results, err := c.client.CreateTransfers(transfers)
	if err != nil {
		return &TransferResult{
			TransferID:   transferID,
			Status:       "failed",
			ErrorMessage: err.Error(),
			Timestamp:    time.Now(),
		}, fmt.Errorf("failed to void pending transfer: %w", err)
	}

	if len(results) > 0 {
		return &TransferResult{
			TransferID:   transferID,
			Status:       "failed",
			ErrorCode:    fmt.Sprintf("%d", results[0].Result),
			ErrorMessage: "Void pending transfer failed",
			Timestamp:    time.Now(),
		}, fmt.Errorf("void pending transfer failed with result: %v", results[0].Result)
	}

	return &TransferResult{
		TransferID: transferID,
		Status:     "voided",
		Timestamp:  time.Now(),
	}, nil
}

func (c *TigerBeetleClient) GetAccountBalance(ctx context.Context, accountID types.Uint128) (*AccountBalance, error) {
	accounts, err := c.client.LookupAccounts([]types.Uint128{accountID})
	if err != nil {
		return nil, fmt.Errorf("failed to lookup account: %w", err)
	}

	if len(accounts) == 0 {
		return nil, fmt.Errorf("account not found")
	}

	account := accounts[0]
	netBalance := int64(account.CreditsPosted) - int64(account.DebitsPosted)

	return &AccountBalance{
		AccountID:      accountID,
		DebitsPosted:   account.DebitsPosted,
		CreditsPosted:  account.CreditsPosted,
		DebitsPending:  account.DebitsPending,
		CreditsPending: account.CreditsPending,
		NetBalance:     netBalance,
	}, nil
}

func (c *TigerBeetleClient) Close() {
	c.client.Close()
}

type TigerBeetleActivities struct {
	client *TigerBeetleClient
}

func NewTigerBeetleActivities(client *TigerBeetleClient) *TigerBeetleActivities {
	return &TigerBeetleActivities{
		client: client,
	}
}

func (a *TigerBeetleActivities) CreateAccountActivity(ctx context.Context, accountID types.Uint128, ledger uint32, code uint16) error {
	logger := activity.GetLogger(ctx)
	logger.Info("Creating TigerBeetle account", "accountID", accountID)

	if err := a.client.CreateAccount(ctx, accountID, ledger, code); err != nil {
		logger.Error("Failed to create account", "error", err)
		return err
	}

	logger.Info("Account created successfully", "accountID", accountID)
	return nil
}

func (a *TigerBeetleActivities) CreateTransferActivity(ctx context.Context, req TransferRequest) (*TransferResult, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Creating TigerBeetle transfer", "transferID", req.TransferID)

	result, err := a.client.CreateTransfer(ctx, req)
	if err != nil {
		logger.Error("Failed to create transfer", "error", err)
		return result, err
	}

	logger.Info("Transfer created successfully", "transferID", req.TransferID, "status", result.Status)
	return result, nil
}

func (a *TigerBeetleActivities) PostPendingTransferActivity(ctx context.Context, transferID, pendingTransferID string, ledger uint32, code uint16) (*TransferResult, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Posting pending transfer", "transferID", transferID, "pendingTransferID", pendingTransferID)

	result, err := a.client.PostPendingTransfer(ctx, transferID, pendingTransferID, ledger, code)
	if err != nil {
		logger.Error("Failed to post pending transfer", "error", err)
		return result, err
	}

	logger.Info("Pending transfer posted successfully", "transferID", transferID)
	return result, nil
}

func (a *TigerBeetleActivities) VoidPendingTransferActivity(ctx context.Context, transferID, pendingTransferID string, ledger uint32, code uint16) (*TransferResult, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Voiding pending transfer", "transferID", transferID, "pendingTransferID", pendingTransferID)

	result, err := a.client.VoidPendingTransfer(ctx, transferID, pendingTransferID, ledger, code)
	if err != nil {
		logger.Error("Failed to void pending transfer", "error", err)
		return result, err
	}

	logger.Info("Pending transfer voided successfully", "transferID", transferID)
	return result, nil
}

func (a *TigerBeetleActivities) GetAccountBalanceActivity(ctx context.Context, accountID types.Uint128) (*AccountBalance, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Getting account balance", "accountID", accountID)

	balance, err := a.client.GetAccountBalance(ctx, accountID)
	if err != nil {
		logger.Error("Failed to get account balance", "error", err)
		return nil, err
	}

	logger.Info("Account balance retrieved", "accountID", accountID, "netBalance", balance.NetBalance)
	return balance, nil
}

type PaymentWorkflowInput struct {
	PaymentID       string
	DebitAccountID  types.Uint128
	CreditAccountID types.Uint128
	Amount          uint64
	Currency        string
	Ledger          uint32
	Code            uint16
}

type PaymentWorkflowResult struct {
	PaymentID    string
	TransferID   string
	Status       string
	ErrorMessage string
	CompletedAt  time.Time
}

func PaymentWorkflow(ctx workflow.Context, input PaymentWorkflowInput) (*PaymentWorkflowResult, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("Starting payment workflow", "paymentID", input.PaymentID)

	ao := workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
		RetryPolicy: &workflow.RetryPolicy{
			InitialInterval:    time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    time.Minute,
			MaximumAttempts:    3,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, ao)

	transferID := fmt.Sprintf("TXN-%s-%d", input.PaymentID, time.Now().Unix())

	pendingReq := TransferRequest{
		TransferID:      transferID,
		DebitAccountID:  input.DebitAccountID,
		CreditAccountID: input.CreditAccountID,
		Amount:          input.Amount,
		Ledger:          input.Ledger,
		Code:            input.Code,
		Timeout:         3600,
		IsPending:       true,
	}

	var pendingResult *TransferResult
	err := workflow.ExecuteActivity(ctx, "CreateTransferActivity", pendingReq).Get(ctx, &pendingResult)
	if err != nil {
		logger.Error("Failed to create pending transfer", "error", err)
		return &PaymentWorkflowResult{
			PaymentID:    input.PaymentID,
			Status:       "failed",
			ErrorMessage: err.Error(),
			CompletedAt:  time.Now(),
		}, err
	}

	var paymentApproved bool
	err = workflow.ExecuteActivity(ctx, "ValidatePaymentActivity", input.PaymentID).Get(ctx, &paymentApproved)
	if err != nil {
		logger.Error("Payment validation failed", "error", err)
		
		voidID := fmt.Sprintf("VOID-%s", transferID)
		var voidResult *TransferResult
		workflow.ExecuteActivity(ctx, "VoidPendingTransferActivity", voidID, transferID, input.Ledger, input.Code).Get(ctx, &voidResult)
		
		return &PaymentWorkflowResult{
			PaymentID:    input.PaymentID,
			Status:       "failed",
			ErrorMessage: "Payment validation failed",
			CompletedAt:  time.Now(),
		}, err
	}

	if !paymentApproved {
		logger.Info("Payment not approved, voiding transfer")
		
		voidID := fmt.Sprintf("VOID-%s", transferID)
		var voidResult *TransferResult
		err = workflow.ExecuteActivity(ctx, "VoidPendingTransferActivity", voidID, transferID, input.Ledger, input.Code).Get(ctx, &voidResult)
		if err != nil {
			logger.Error("Failed to void pending transfer", "error", err)
		}
		
		return &PaymentWorkflowResult{
			PaymentID:    input.PaymentID,
			TransferID:   transferID,
			Status:       "rejected",
			ErrorMessage: "Payment not approved",
			CompletedAt:  time.Now(),
		}, nil
	}

	postID := fmt.Sprintf("POST-%s", transferID)
	var postResult *TransferResult
	err = workflow.ExecuteActivity(ctx, "PostPendingTransferActivity", postID, transferID, input.Ledger, input.Code).Get(ctx, &postResult)
	if err != nil {
		logger.Error("Failed to post pending transfer", "error", err)
		return &PaymentWorkflowResult{
			PaymentID:    input.PaymentID,
			TransferID:   transferID,
			Status:       "failed",
			ErrorMessage: err.Error(),
			CompletedAt:  time.Now(),
		}, err
	}

	logger.Info("Payment workflow completed successfully", "paymentID", input.PaymentID)
	return &PaymentWorkflowResult{
		PaymentID:   input.PaymentID,
		TransferID:  transferID,
		Status:      "completed",
		CompletedAt: time.Now(),
	}, nil
}

func ClaimPaymentWorkflow(ctx workflow.Context, input PaymentWorkflowInput) (*PaymentWorkflowResult, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("Starting claim payment workflow", "paymentID", input.PaymentID)

	ao := workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
		RetryPolicy: &workflow.RetryPolicy{
			InitialInterval:    time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    time.Minute,
			MaximumAttempts:    3,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, ao)

	var claimApproved bool
	err := workflow.ExecuteActivity(ctx, "ValidateClaimActivity", input.PaymentID).Get(ctx, &claimApproved)
	if err != nil || !claimApproved {
		logger.Error("Claim validation failed", "error", err)
		return &PaymentWorkflowResult{
			PaymentID:    input.PaymentID,
			Status:       "rejected",
			ErrorMessage: "Claim validation failed",
			CompletedAt:  time.Now(),
		}, err
	}

	transferID := fmt.Sprintf("CLAIM-%s-%d", input.PaymentID, time.Now().Unix())

	transferReq := TransferRequest{
		TransferID:      transferID,
		DebitAccountID:  input.DebitAccountID,
		CreditAccountID: input.CreditAccountID,
		Amount:          input.Amount,
		Ledger:          input.Ledger,
		Code:            input.Code,
		IsPending:       false,
	}

	var transferResult *TransferResult
	err = workflow.ExecuteActivity(ctx, "CreateTransferActivity", transferReq).Get(ctx, &transferResult)
	if err != nil {
		logger.Error("Failed to create claim transfer", "error", err)
		return &PaymentWorkflowResult{
			PaymentID:    input.PaymentID,
			Status:       "failed",
			ErrorMessage: err.Error(),
			CompletedAt:  time.Now(),
		}, err
	}

	logger.Info("Claim payment workflow completed successfully", "paymentID", input.PaymentID)
	return &PaymentWorkflowResult{
		PaymentID:   input.PaymentID,
		TransferID:  transferID,
		Status:      "completed",
		CompletedAt: time.Now(),
	}, nil
}

func parseTransferID(transferID string) types.Uint128 {
	id := uuid.MustParse(transferID)
	high := uint64(id[0])<<56 | uint64(id[1])<<48 | uint64(id[2])<<40 | uint64(id[3])<<32 |
		uint64(id[4])<<24 | uint64(id[5])<<16 | uint64(id[6])<<8 | uint64(id[7])
	low := uint64(id[8])<<56 | uint64(id[9])<<48 | uint64(id[10])<<40 | uint64(id[11])<<32 |
		uint64(id[12])<<24 | uint64(id[13])<<16 | uint64(id[14])<<8 | uint64(id[15])

	return types.Uint128{
		High: high,
		Low:  low,
	}
}
